File: macos/Sources/CursorTokenCore/CloudSync.swift
import Foundation

public enum CloudSyncApi {
    public static let baseUrl = "https://sync.harker.cn"
}

public enum CloudSync {
    public static func applySession(_ cfg: inout AppConfig, email: String, password: String, access: String, refresh: String) {
        cfg.cloudEmail = email.trimmingCharacters(in: .whitespaces).lowercased()
        cfg.cloudAccessToken = access
        cfg.cloudRefreshToken = refresh
        cfg.syncSecret = password
        cfg.syncEnabled = cfg.cloudLoggedIn
        cfg.syncLastError = ""
        cfg.syncSecretDecryptFailed = false
    }

    public static func clearSession(_ cfg: inout AppConfig, keepEmail: Bool = true) {
        if !keepEmail { cfg.cloudEmail = "" }
        cfg.cloudAccessToken = ""
        cfg.cloudRefreshToken = ""
        cfg.syncSecret = ""
        cfg.cloudRevision = 0
        cfg.syncEnabled = false
    }

    public static func register(email: String, password: String) throws -> (email: String, access: String, refresh: String) {
        let payload = try send(method: "POST", path: "/v1/auth/register", access: nil, body: ["email": email, "password": password])
        return try readAuth(payload)
    }

    public static func login(email: String, password: String) throws -> (email: String, access: String, refresh: String) {
        let payload = try send(method: "POST", path: "/v1/auth/login", access: nil, body: ["email": email, "password": password])
        return try readAuth(payload)
    }

    public static func logout(_ cfg: inout AppConfig) {
        let refresh = cfg.cloudRefreshToken
        if !refresh.isEmpty {
            _ = try? send(method: "POST", path: "/v1/auth/logout", access: nil, body: ["refresh_token": refresh])
        }
        clearSession(&cfg)
    }

    public static func reconcile(_ cfg: inout AppConfig, now: Date? = nil, write: Bool = true) -> SyncStatus {
        var status = SyncStatus(path: CloudSyncApi.baseUrl)
        let reason = AccountSync.syncReady(cfg)
        if !reason.isEmpty {
            status.message = reason
            cfg.syncLastError = reason
            return status
        }
        _ = AccountSync.ensureDeviceId(&cfg)
        let stamp = AccountSync.nowIso(now)
        let passphrase = cfg.syncSecret
        var local = AccountSync.snapshotFromConfig(cfg)
        local.deviceId = cfg.syncDeviceId
        do {
            let got = try authed(&cfg, method: "GET", path: "/v1/sync", body: nil)
            let revision = intValue(got["revision"]) ?? 0
            let remote = try parseRemote(got, passphrase: passphrase)
            let merged = AccountSync.mergeSnapshots(local, remote)
            let changed = AccountSync.applySnapshotToConfig(&cfg, merged)
            if write && AccountSync.snapshotIdentity(merged) != AccountSync.snapshotIdentity(remote) {
                var toWrite = merged
                toWrite.updatedAt = stamp
                toWrite.deviceId = cfg.syncDeviceId
                let envelope = try AccountSync.encryptEnvelope(toWrite, passphrase: passphrase)
                let put = try authed(&cfg, method: "PUT", path: "/v1/sync", body: ["revision": revision, "envelope": envelope])
                cfg.cloudRevision = intValue(put["revision"]) ?? (revision + 1)
                status.pushed = true
            } else {
                cfg.cloudRevision = revision
            }
            cfg.syncLastAt = stamp
            cfg.syncLastError = ""
            status.ok = true
            status.changed = changed
            if changed && status.pushed { status.message = "已合并并对齐云端" }
            else if changed { status.message = "已从云端导入账号和设置" }
            else if status.pushed { status.message = "已上传到云端" }
            else { status.message = "账号和设置已与云端一致" }
            return status
        } catch let err as CursorAPIError where err.statusCode == 401 {
            clearSession(&cfg)
            status.message = "登录已过期，请重新登录"
            cfg.syncLastError = status.message
            return status
        } catch {
            let message = (error as? CursorAPIError)?.message ?? error.localizedDescription
            cfg.syncLastError = message.isEmpty ? "同步失败" : message
            status.message = cfg.syncLastError
            return status
        }
    }

    static func parseRemote(_ payload: [String: Any], passphrase: String) throws -> SyncSnapshot {
        guard let envelope = payload["envelope"] as? [String: Any] else { return SyncSnapshot() }
        return try AccountSync.decryptEnvelope(envelope, passphrase: passphrase)
    }

    static func readAuth(_ payload: [String: Any]) throws -> (email: String, access: String, refresh: String) {
        let access = str(payload["access_token"])
        let refresh = str(payload["refresh_token"])
        let email = str(payload["email"])
        if access.isEmpty { throw CursorAPIError("登录失败") }
        return (email, access, refresh)
    }

    static func authed(_ cfg: inout AppConfig, method: String, path: String, body: [String: Any]?) throws -> [String: Any] {
        do {
            return try send(method: method, path: path, access: cfg.cloudAccessToken, body: body)
        } catch let err as CursorAPIError where err.statusCode == 401 {
            if !refresh(&cfg) { throw err }
            return try send(method: method, path: path, access: cfg.cloudAccessToken, body: body)
        }
    }

    static func refresh(_ cfg: inout AppConfig) -> Bool {
        let token = cfg.cloudRefreshToken.trimmingCharacters(in: .whitespaces)
        if token.isEmpty { return false }
        do {
            let payload = try send(method: "POST", path: "/v1/auth/refresh", access: nil, body: ["refresh_token": token])
            cfg.cloudAccessToken = str(payload["access_token"])
            if let next = payload["refresh_token"] as? String, !next.isEmpty {
                cfg.cloudRefreshToken = next
            }
            return !cfg.cloudAccessToken.isEmpty
        } catch {
            clearSession(&cfg)
            return false
        }
    }

    static func send(method: String, path: String, access: String?, body: [String: Any]?) throws -> [String: Any] {
        guard let url = URL(string: CloudSyncApi.baseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + path) else {
            throw CursorAPIError("无法连接同步服务器")
        }
        var req = URLRequest(url: url, timeoutInterval: 20)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        if let access, !access.isEmpty {
            req.setValue("Bearer \(access)", forHTTPHeaderField: "Authorization")
        }
        if let body {
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
        }
        let box = RequestBox()
        let sem = DispatchSemaphore(value: 0)
        URLSession.shared.dataTask(with: req) { data, response, error in
            box.data = data
            box.response = response
            box.error = error
            sem.signal()
        }.resume()
        _ = sem.wait(timeout: .now() + 25)
        if box.error != nil { throw CursorAPIError("无法连接同步服务器") }
        let status = (box.response as? HTTPURLResponse)?.statusCode ?? 0
        let obj = (try? JSONSerialization.jsonObject(with: box.data ?? Data())) as? [String: Any] ?? [:]
        if status >= 400 {
            let detail = str(obj["detail"])
            throw CursorAPIError(detail.isEmpty ? "同步失败" : detail, statusCode: status)
        }
        return obj
    }

    static func str(_ value: Any?) -> String {
        (value as? String) ?? ""
    }

    static func intValue(_ value: Any?) -> Int? {
        if let n = value as? Int { return n }
        if let n = value as? NSNumber { return n.intValue }
        return nil
    }
}

private final class RequestBox: @unchecked Sendable {
    var data: Data?
    var response: URLResponse?
    var error: Error?
}

File: windows/CursorTokenCore/CloudSync.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace CursorTokenCore;

public static class CloudSyncApi
{
    public const string BaseUrl = "https://sync.harker.cn";
}

public static class CloudSync
{
    static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(20) };
    static readonly JsonSerializerOptions JsonOpt = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
        PropertyNameCaseInsensitive = true,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    public static void ApplySession(AppConfig cfg, string email, string password, string access, string refresh)
    {
        cfg.CloudEmail = (email ?? "").Trim().ToLowerInvariant();
        cfg.CloudAccessToken = access ?? "";
        cfg.CloudRefreshToken = refresh ?? "";
        cfg.SyncSecret = password ?? "";
        cfg.SyncEnabled = cfg.CloudLoggedIn;
        cfg.SyncLastError = "";
        cfg.SyncSecretDecryptFailed = false;
    }

    public static void ClearSession(AppConfig cfg, bool keepEmail = true)
    {
        if (!keepEmail) cfg.CloudEmail = "";
        cfg.CloudAccessToken = "";
        cfg.CloudRefreshToken = "";
        cfg.SyncSecret = "";
        cfg.CloudRevision = 0;
        cfg.SyncEnabled = false;
    }

    public static SyncStatus Reconcile(AppConfig cfg, DateTimeOffset? now = null, bool write = true) =>
        ReconcileAsync(cfg, now, write).GetAwaiter().GetResult();

    public static async Task<(string Email, string Access, string Refresh)> RegisterAsync(string email, string password, CancellationToken ct = default)
    {
        var payload = await SendAsync("POST", "/v1/auth/register", null, new { email, password }, ct);
        return ReadAuth(payload);
    }

    public static async Task<(string Email, string Access, string Refresh)> LoginAsync(string email, string password, CancellationToken ct = default)
    {
        var payload = await SendAsync("POST", "/v1/auth/login", null, new { email, password }, ct);
        return ReadAuth(payload);
    }

    public static async Task LogoutAsync(AppConfig cfg, CancellationToken ct = default)
    {
        var refresh = cfg.CloudRefreshToken;
        if (!string.IsNullOrWhiteSpace(refresh))
        {
            try { await SendAsync("POST", "/v1/auth/logout", null, new { refresh_token = refresh }, ct); }
            catch { }
        }
        ClearSession(cfg);
    }

    public static async Task<SyncStatus> ReconcileAsync(AppConfig cfg, DateTimeOffset? now = null, bool write = true, CancellationToken ct = default)
    {
        var status = new SyncStatus { Path = CloudSyncApi.BaseUrl };
        var reason = AccountSync.SyncReady(cfg);
        if (reason.Length > 0)
        {
            status.Message = reason;
            cfg.SyncLastError = reason;
            return status;
        }
        AccountSync.EnsureDeviceId(cfg);
        var stamp = AccountSync.NowIso(now);
        var passphrase = cfg.SyncSecret;
        var local = AccountSync.SnapshotFromConfig(cfg);
        local.DeviceId = cfg.SyncDeviceId;
        try
        {
            using var got = await AuthedAsync(cfg, HttpMethod.Get, "/v1/sync", null, ct);
            var revision = got.RootElement.TryGetProperty("revision", out var rev) && rev.TryGetInt32(out var n) ? n : 0;
            var remote = ParseRemote(got.RootElement, passphrase);
            var merged = AccountSync.MergeSnapshots(local, remote);
            var changed = AccountSync.ApplySnapshotToConfig(cfg, merged);
            if (write && AccountSync.SnapshotIdentity(merged) != AccountSync.SnapshotIdentity(remote))
            {
                merged.UpdatedAt = stamp;
                merged.DeviceId = cfg.SyncDeviceId;
                var envelope = AccountSync.EncryptEnvelope(merged, passphrase);
                using var put = await AuthedAsync(cfg, HttpMethod.Put, "/v1/sync", new { revision, envelope }, ct);
                if (put.RootElement.TryGetProperty("revision", out var next) && next.TryGetInt32(out var nextRev))
                    cfg.CloudRevision = nextRev;
                else
                    cfg.CloudRevision = revision + 1;
                status.Pushed = true;
            }
            else cfg.CloudRevision = revision;
            cfg.SyncLastAt = stamp;
            cfg.SyncLastError = "";
            status.Ok = true;
            status.Changed = changed;
            status.Message = changed && status.Pushed ? "已合并并对齐云端"
                : changed ? "已从云端导入账号和设置"
                : status.Pushed ? "已上传到云端"
                : "账号和设置已与云端一致";
            return status;
        }
        catch (CursorApiException ex) when (ex.StatusCode == 401)
        {
            ClearSession(cfg);
            status.Message = "登录已过期，请重新登录";
            cfg.SyncLastError = status.Message;
            return status;
        }
        catch (Exception ex)
        {
            var message = string.IsNullOrWhiteSpace(ex.Message) ? "同步失败" : ex.Message;
            cfg.SyncLastError = message;
            status.Message = message;
            return status;
        }
    }

    static SyncSnapshot ParseRemote(JsonElement root, string passphrase)
    {
        if (!root.TryGetProperty("envelope", out var env) || env.ValueKind is JsonValueKind.Null or JsonValueKind.Undefined)
            return new SyncSnapshot();
        return AccountSync.DecryptEnvelope(env, passphrase);
    }

    static (string Email, string Access, string Refresh) ReadAuth(JsonDocument payload)
    {
        var root = payload.RootElement;
        var access = root.TryGetProperty("access_token", out var a) ? a.GetString() ?? "" : "";
        var refresh = root.TryGetProperty("refresh_token", out var r) ? r.GetString() ?? "" : "";
        var email = root.TryGetProperty("email", out var e) ? e.GetString() ?? "" : "";
        if (access.Length == 0) throw new CursorApiException("登录失败");
        return (email, access, refresh);
    }

    static async Task<JsonDocument> AuthedAsync(AppConfig cfg, HttpMethod method, string path, object? body, CancellationToken ct)
    {
        try
        {
            return await SendAsync(method.Method, path, cfg.CloudAccessToken, body, ct);
        }
        catch (CursorApiException ex) when (ex.StatusCode == 401)
        {
            if (!await RefreshAsync(cfg, ct)) throw;
            return await SendAsync(method.Method, path, cfg.CloudAccessToken, body, ct);
        }
    }

    static async Task<bool> RefreshAsync(AppConfig cfg, CancellationToken ct)
    {
        var refresh = (cfg.CloudRefreshToken ?? "").Trim();
        if (refresh.Length == 0) return false;
        try
        {
            using var doc = await SendAsync("POST", "/v1/auth/refresh", null, new { refresh_token = refresh }, ct);
            var root = doc.RootElement;
            cfg.CloudAccessToken = root.TryGetProperty("access_token", out var a) ? a.GetString() ?? "" : "";
            if (root.TryGetProperty("refresh_token", out var r))
                cfg.CloudRefreshToken = r.GetString() ?? cfg.CloudRefreshToken ?? "";
            return cfg.CloudAccessToken.Length > 0;
        }
        catch
        {
            ClearSession(cfg);
            return false;
        }
    }

    static async Task<JsonDocument> SendAsync(string method, string path, string? access, object? body, CancellationToken ct)
    {
        using var req = new HttpRequestMessage(new HttpMethod(method), CloudSyncApi.BaseUrl.TrimEnd('/') + path);
        req.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        if (!string.IsNullOrWhiteSpace(access))
            req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", access);
        if (body is not null)
            req.Content = new StringContent(JsonSerializer.Serialize(body, JsonOpt), Encoding.UTF8, "application/json");
        HttpResponseMessage res;
        try
        {
            res = await Http.SendAsync(req, ct);
        }
        catch (OperationCanceledException) when (!ct.IsCancellationRequested)
        {
            throw new CursorApiException("无法连接同步服务器");
        }
        catch (HttpRequestException)
        {
            throw new CursorApiException("无法连接同步服务器");
        }
        using (res)
        {
        var text = await res.Content.ReadAsStringAsync(ct);
        JsonDocument doc;
        try { doc = JsonDocument.Parse(string.IsNullOrWhiteSpace(text) ? "{}" : text); }
        catch { doc = JsonDocument.Parse("{}"); }
        if ((int)res.StatusCode >= 400)
        {
            var detail = doc.RootElement.TryGetProperty("detail", out var d) && d.ValueKind == JsonValueKind.String
                ? d.GetString()
                : null;
            throw new CursorApiException(string.IsNullOrWhiteSpace(detail) ? "同步失败" : detail!, (int)res.StatusCode);
        }
        return doc;
        }
    }
}
